
* *Name:* Neaya Chalise
* Roll Number: 11

# Data Preprocessing

## Project Objectives

### 1. Data Cleaning
- Identify data quality issues.
- Handle missing values.
- Fix inconsistent formats and invalid entries.

### 2. Data Preprocessing
- Standardize data types.
- Apply necessary transformation techniques.
- Prepare clean and structured data for analysis.

### 3. Optimization
- Optimize the processed dataset for:
  - Exploratory Data Analysis (EDA)
  - Data Visualization
  - Predictive Modeling



## Dataset Description

The dataset contains information about products listed on Amazon along with their pricing, ratings, and categorical details.
Each row represents a single product.

### Columns Description:
- **product_id**: Unique identifier for each product.
- **product_name**: Name/title of the product as listed on Amazon.
- **category**: Product category hierarchy separated by `|`.
- **discounted_price**: Selling price after discount (originally in ₹ format).
- **actual_price**: Original price before discount (₹ format).
- **discount_percentage**: Discount offered on the product (percentage).
- **rating**: Average customer rating of the product.
- **rating_count**: Total number of ratings received.
- **about_product**: Brief description of the product.
- **user_id**: Unique identifier of the reviewer.
- **user_name**: Name of the reviewer.
- **review_id**: Unique review identifier.
- **review_title**: Title of the review.
- **review_content**: Detailed customer review.
- **img_link**: URL of the product image.
- **product_link**: URL of the product page on Amazon.

In [450]:
# Importing necessary libraries
import pandas as pd
import numpy as np

In [451]:
# Load the Amazon dataset from CSV file
df = pd.read_csv('amazon.csv')
# Displaying the first few rows of the dataset
df.head()

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [452]:
# Displaying the Last few rows of the dataset
df.tail()

product_id                                       product_name  \
1460  B08L7J3T31  Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...   
1461  B01M6453MB  Prestige Delight PRWO Electric Rice Cooker (1 ...   
1462  B009P2LIL4  Bajaj Majesty RX10 2000 Watts Heat Convector R...   
1463  B00J5DYCCA  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
1464  B01486F4G6  Borosil Jumbo 1000-Watt Grill Sandwich Maker (...   

                                               category discounted_price  \
1460  Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...             ₹379   
1461  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,280   
1462  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...           ₹2,219   
1463  Home&Kitchen|Heating,Cooling&AirQuality|Fans|E...           ₹1,399   
1464  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,863   

     actual_price discount_percentage rating rating_count  \
1460         ₹919                 59%      4        1,090   
1461       ₹3,045                 25%    4.1        4,118   
1462       ₹3,080                 28%    3.6          468   
1463       ₹1,890                 26%      4        8,031   
1464       ₹3,690                 22%    4.3        6,987   

                                          about_product  \
1460  SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...   
1461                       230 Volts, 400 watts, 1 Year   
1462  International design and styling|Two heat sett...   
1463  Fan sweep area: 230 MM ; Noise level: (40 - 45...   
1464  Brand-Borosil, Specification – 23V ~ 5Hz;1 W...   

                                                user_id  \
1460  AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...   
1461  AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...   
1462  AGVPWCMAHYQWJOQKMUJN4DW3KM5Q,AF4Q3E66MY4SR7YQZ...   
1463  AF2JQCLSCY3QJATWUNNHUSVUPNQQ,AFDMLUXC5LS5RXDJS...   
1464  AFGW5PT3R6ZAVQR4Y5MWVAKBZAYA,AG7QNJ2SCS5VS5VYY...   

                                              user_name  \
1460  Prabha ds,Raghuram bk,Real Deal,Amazon Custome...   
1461  Manu Bhai,Naveenpittu,Evatira Sangma,JAGANNADH...   
1462  Nehal Desai,Danish Parwez,Amazon Customer,Amaz...   
1463  Shubham Dubey,E.GURUBARAN,Mayank S.,eusuf khan...   
1464  Rajib,Ajay B,Vikas Kahol,PARDEEP,Anindya Prama...   

                                              review_id  \
1460  R3G3XFHPBFF0E8,R3C0BZCD32EIGW,R2EBVBCN9QPD9R,R...   
1461  R3DDL2UPKQ2CK9,R2SYYU1OATVIU5,R1VM993161IYRW,R...   
1462  R1TLRJVW4STY5I,R2O455KRN493R1,R3Q5MVGBRIAS2G,R...   
1463  R39Q2Y79MM9SWK,R3079BG1NIH6MB,R29A31ZELTZNJM,R...   
1464  R20RBRZ0WEUJT9,ROKIFK9R2ISSE,R30EEG2FNJSN5I,R2...   

                                           review_title  \
1460  Received the product without spanner,Excellent...   
1461  ok,everything was good couldn't return bcoz I ...   
1462  very good,Work but front melt after 2 month,Go...   
1463  Fan Speed is slow,Good quality,Good product,go...   
1464  Works perfect,Ok good product,Nice Product. Re...   

                                         review_content  \
1460  I received product without spanner,Excellent p...   
1461  ok,got everything as mentioned but the measuri...   
1462  plastic but cool body ,u have to find sturdy s...   
1463  I have installed this in my kitchen working fi...   
1464  It does it job perfectly..only issue is temp c...   

                                               img_link  \
1460  https://m.media-amazon.com/images/I/41fDdRtjfx...   
1461  https://m.media-amazon.com/images/I/41gzDxk4+k...   
1462  https://m.media-amazon.com/images/W/WEBP_40237...   
1463  https://m.media-amazon.com/images/W/WEBP_40237...   
1464  https://m.media-amazon.com/images/W/WEBP_40237...   

                                           product_link  
1460  https://www.amazon.in/Noir-Aqua-Spanner-Purifi...  
1461  https://www.amazon.in/Prestige-Delight-PRWO-1-...  
1462  https://www.amazon.in/Bajaj-RX-10-2000-Watt-Co...  
1463  https://www.amazon.in/Havells-Ve

## Initial Data Inspection

In [453]:
# Checking the shape of the dataset (number of rows and columns)
print("Dataset shape: ")
print(df.shape)


Dataset shape: 
(1465, 16)


In [454]:
# Getting information about the dataset (data types, non-null counts, etc.)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

In [455]:

# Getting summary statistics of the dataset
df.describe()

product_id                                       product_name  \
count         1465                                               1465   
unique        1351                                               1337   
top     B08WRWPM22  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
freq             3                                                  5   

                                                 category discounted_price  \
count                                                1465             1465   
unique                                                211              550   
top     Computers&Accessories|Accessories&Peripherals|...             ₹199   
freq                                                  233               53   

       actual_price discount_percentage rating rating_count  \
count          1465                1465   1465         1463   
unique          449                  92     28         1143   
top            ₹999                 50%    4.1        9,378   
freq            120                  56    244            9   

                                            about_product  \
count                                                1465   
unique                                               1293   
top     [CHARGE & SYNC FUNCTION]- This cable comes wit...   
freq                                                    6   

                                                  user_id  \
count                                                1465   
unique                                               1194   
top     AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...   
freq                                                   10   

                                                user_name  \
count                                                1465   
unique                                               1194   
top     $@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...   
freq                                                   10   

                                                review_id  \
count                                                1465   
unique                                               1194   
top     R3F4T5TRYPTMIG,R3DQIEC603E7AY,R1O4Z15FD40PV5,R...   
freq                                                   10   

                                             review_title  \
count                                                1465   
unique                                               1194   
top     Worked on iPhone 7 and didn’t work on XR,Good ...   
freq                                                   10   

                                           review_content  \
count                                                1465   
unique                                               1212   
top     I am not big on camera usage, personally. I wa...   
freq                                                    8   

                                                 img_link  \
count                                                1465   
unique                                               1412   
top     https://m.media-amazon.com/images/I/413sCRKobN...   
freq                                                    3   

                                             product_link  
count                                                1465  
unique                                               1465  
top     https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
freq                                                    1

In [456]:
# Check the type of the object (DataFrame or Series)
type(df)

pandas.core.frame.DataFrame

#### **Dataset Dimensions and Summary**

* **Shape Analysis:** The `shape` attribute is used to determine the total number of rows and columns in the dataset.
* **Data Metadata:** The `info()` method helps identify data types and provides a quick count of non-null values to pinpoint missing data.
* **Statistical Summary:** The `describe()` function generates summary statistics (mean, median, min, max, etc.) for numerical columns to understand data distribution.

## Handling the missing values

In [457]:
# Checking for missing values in the dataset
df.isnull().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           2
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64

In [458]:
# Remove commas from rating_count (string manipulation)
df['rating_count'] = df['rating_count'].str.replace(',', '')

In [459]:
# Convert to nullable integer type (Int64) and if we just use int instead of Int64 we
# will get an error as it cannot change null value to int
df['rating_count'] = df['rating_count'].astype('Int64')

In [460]:
# Fill missing values in rating_count using median
df['rating_count'] = df['rating_count'].fillna(df['rating_count'].median())

In [461]:
# Check the number of missing values in each column
df.isnull().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           0
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64

In [462]:
# Remove unwanted '|' symbols from rating column
df['rating'] = df['rating'].str.replace('|', '',regex=False)

In [463]:
# Replace empty or blank values with NaN
df['rating'] = df['rating'].replace(r'^\s*$', np.nan, regex=True)

In [464]:
# Convert rating to float
df['rating'] = df['rating'].astype(float)

#### Handling Missing Values

Missing values were identified in the `rating_count` column. Since this feature represents user engagement and may be skewed, missing values were imputed using the median to preserve the distribution and reduce the influence of outliers.


## Price Cleaning

In [465]:
# Remove rupee symbol from discounted price
df['discounted_price'] = df['discounted_price'].str.replace('₹', '', regex=False)

In [466]:
# Remove commas from discounted price
df['discounted_price'] = df['discounted_price'].str.replace(',', '', regex=False)

In [467]:
# Convert discounted price to float
df['discounted_price'] = df['discounted_price'].astype(float)

In [468]:
# Remove percentage symbol and convert discount percentage to float
df['discount_percentage'] = (
    df['discount_percentage']
    .str.replace('%', '', regex=False)
    .astype(float)
)

In [469]:
# Remove rupee symbol from actual price
df['actual_price'] = df['actual_price'].str.replace('₹', '', regex=False)

In [470]:
# Remove commas from actual price
df['actual_price'] = df['actual_price'].str.replace(',', '', regex=False)

In [471]:
# Convert actual price to float
df['actual_price'] = df['actual_price'].astype(float)

In [472]:
#check if the actual price is negative
negative_prices = df['actual_price'] < 0
negative_prices


0       False
1       False
2       False
3       False
4       False
        ...  
1460    False
1461    False
1462    False
1463    False
1464    False
Name: actual_price, Length: 1465, dtype: bool

In [473]:
#check if the discounted price is negative
negative_prices1 = df['discounted_price'] < 0
negative_prices1

0       False
1       False
2       False
3       False
4       False
        ...  
1460    False
1461    False
1462    False
1463    False
1464    False
Name: discounted_price, Length: 1465, dtype: bool

#### Category Splitting

The `category` column contains hierarchical information separated by a delimiter. This column was split to extract the `main_category`, which simplifies grouping, aggregation, and category-level analysis.


## Feature Engineering

In [474]:
# Create a new feature: discounted amount (price reduction)
df['discounted_amount'] = df['actual_price'] - df['discounted_price']
# Display the first few rows of the updated dataset
df.head()

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category  discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             399.0   
1  Computers&Accessories|Accessories&Peripherals|...             199.0   
2  Computers&Accessories|Accessories&Peripherals|...             199.0   
3  Computers&Accessories|Accessories&Peripherals|...             329.0   
4  Computers&Accessories|Accessories&Peripherals|...             154.0   

   actual_price  discount_percentage  rating  rating_count  \
0        1099.0                 64.0     4.2         24269   
1         349.0                 43.0     4.0         43994   
2        1899.0                 90.0     3.9          7928   
3         699.0                 53.0     4.2         94363   
4         399.0                 61.0     4.2         16905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  discounted_amount  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...              700.0  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...              150.0  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...             1700.0  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...              370.0  
4  https://www.amazon.in/Portronics-Konnect-POR

## Threshold-based feature engineering

In [475]:
# Create a binary flag for highly rated products (1 if the rating is above or equal to 4 and 0 if it is less than 4)
df['high_rating_flag'] = df['rating'].apply(
    lambda x: 1 if x >= 4 else 0
)
# Display the first few rows of the updated dataset
df.head()

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category  discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             399.0   
1  Computers&Accessories|Accessories&Peripherals|...             199.0   
2  Computers&Accessories|Accessories&Peripherals|...             199.0   
3  Computers&Accessories|Accessories&Peripherals|...             329.0   
4  Computers&Accessories|Accessories&Peripherals|...             154.0   

   actual_price  discount_percentage  rating  rating_count  \
0        1099.0                 64.0     4.2         24269   
1         349.0                 43.0     4.0         43994   
2        1899.0                 90.0     3.9          7928   
3         699.0                 53.0     4.2         94363   
4         399.0                 61.0     4.2         16905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  discounted_amount  \
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...              700.0   
1  https://www.amazon.in/Ambrane-Unbreakable-Char...              150.0   
2  https://www.amazon.in/Sounce-iPhone-Charging-C...             1700.0   
3  https://www.amazon.in/Deuce-300-Resistant-Tang...              370.0   
4  https://www.amazon.in/Portronics-Konnec

#### Feature Engineering

New features were created to enhance analytical value:
- `discount_amount` represents the absolute price reduction.
- `high_rating_flag` identifies products with ratings greater than or equal to 4.

These engineered features provide better insights into pricing strategies and customer satisfaction.


## Outlier Detection

In [476]:
# Calculate first quartile (25%)
Q1 = df['actual_price'].quantile(0.25)

# Calculate third quartile (75%)
Q3 = df['actual_price'].quantile(0.75)


# Calculate Interquartile Range (IQR)
IQR = Q3 - Q1

# Identify outliers using IQR method
outliers = df[
    (df['actual_price'] < Q1 - 1.5 * IQR) |
    (df['actual_price'] > Q3 + 1.5 * IQR)
]
print("Number of outliers in actual_price: ", outliers.shape[0]) 
# Display the outliers
outliers

Number of outliers in actual_price:  213


product_id                                       product_name  \
16    B0B6F7LX4C  MI 80 cm (32 inches) 5A Series HD Ready Smart ...   
19    B08DPLCM6T  LG 80 cm (32 inches) HD Ready Smart LED TV 32L...   
22    B09F6S8BT6  Samsung 80 cm (32 Inches) Wondertainment Serie...   
24    B0B1YVCJ2Y  Acer 80 cm (32 inches) I Series HD Ready Andro...   
26    B08B42LWKN  OnePlus 80 cm (32 inches) Y Series HD Ready LE...   
...          ...                                                ...   
1416  B0BJ6P3LSK  Aqua d pure Active Copper 12-L RO+UV Water Fil...   
1430  B092R48XXB  Mi Robot Vacuum-Mop P, Best-in-class Laser Nav...   
1432  B0977CGNJJ  AGARO Royal Stand 1000W Mixer with 5L SS Bowl ...   
1437  B00JBNZPFM  Karcher WD3 EU Wet and Dry Vacuum Cleaner, 100...   
1438  B08N6P8G5K  INALSA Air Fryer Digital 4L Nutri Fry - 1400W ...   

                                               category  discounted_price  \
16    Electronics|HomeTheater,TV&Video|Televisions|S...           13999.0   
19    Electronics|HomeTheater,TV&Video|Televisions|S...           13490.0   
22    Electronics|HomeTheater,TV&Video|Televisions|S...           13490.0   
24    Electronics|HomeTheater,TV&Video|Televisions|S...           11499.0   
26    Electronics|HomeTheater,TV&Video|Televisions|S...           14999.0   
...                                                 ...               ...   
1416  Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...            4999.0   
1430  Home&Kitchen|Kitchen&HomeAppliances|Vacuum,Cle...           18999.0   
1432  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...            5999.0   
1437  Home&Kitchen|Kitchen&HomeAppliances|Vacuum,Cle...            6199.0   
1438  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...            6790.0   

      actual_price  discount_percentage  rating  rating_count  \
16         24999.0                 44.0     4.2         32840   
19         21990.0                 39.0     4.3         11976   
22         22900.0                 41.0     4.3         16299   
24         19990.0                 42.0     4.3          4703   
26         19999.0                 25.0     4.2         34899   
...            ...                  ...     ...           ...   
1416       24999.0                 80.0     4.5           287   
1430       29999.0                 37.0     4.1          2536   
1432       11495.0                 48.0     4.3           534   
1437       10999.0                 44.0     4.2         10429   
1438       10995.0                 38.0     4.5          3192   

                                          about_product  \
16    Note : The brands, Mi and Xiaomi, are part of ...   
19    Resolution: HD Ready (1366x768) | Refresh Rate...   
22    Resolution: HD Ready (1366x768) | Refresh Rate...   
24    Resolution : HD Ready (1366x768) | Refresh Rat...   
26    Resolution: HD Ready (1366x768) | Refresh Rate...   
...                                                 ...   
1416  Remove 95% to 98% TDS, 100% natural water, Ful...   
1430  Ultimate Power Machine: With a powerful 2100 P...   
1432  1W Motor with 1% copper winding which multi fu...   
1437  Wet & Dry Vacuum Cleaner with patented filter ...   
1438  Oil-free fryer: Fryer with Air Crisp Technolog...   

                                                user_id  \
16    AHEVOQADJSSRX7DS325HSFLMP7VQ,AG7XYZRCSKX6G2OLO...   
19    AHBNKB74LGTYUOKPAJBSKNFV45CA,AFIECTV45ADX5YPTE...   
22    AHEVO4Q5NM4YXMG2HDDXC5XMBGRQ,AFZPH7ZAWX5VDY3HO...   
24    AFSMISGEYDYIP3Z42UTQU4AKOYZQ,AF5ILQY4KFDTO5XHH...   
26    AFUT7ANZTZYGLXU65EQ2D5OP6UMA,AGT7YYJVUC6ZHRKQH...   
...                                                 ...   
1416  AHXO56F7SD2DIP32TF2DYFXQRYLA,AF2JRVSNIBOLEQ7JJ...   
1430  AG33A6XPV67G77FOMXFCNTTPNT4Q,AEOZ6IIJSUVKY2DNO...   
1432  AE33HJDC2ZFONU6UHWESJ4GJ25ZQ,AFFYMSNEPTYBAA3XS...   
1437  AHFX5HMDRZADFXH5XYJLGRDZFM3Q,AEAK4OYOCJC43VRIH...   
1438  AGYYUWOUJKKTMI5CXTJHL6S5YKXQ,AFNA3KTCXDH5PVLFM...   

                             

#### Outlier Detection

Outliers in the `actual_price` column were detected using the Interquartile Range (IQR) method. Identifying outliers helps in understanding extreme pricing behavior and prevents skewed analysis.


## Groupby Operation

In [477]:
# Calculate average rating per category
avg_rating = df.groupby('category')['rating'].mean().reset_index()
#we are using reset_index() because if we donot then category will become index and it will be harder to use for plots or further analysis
# Calculate average discount amount per category
df.groupby('category')['discounted_amount'].mean().reset_index()

category  discounted_amount
0    Car&Motorbike|CarAccessories|InteriorAccessori...        1661.000000
1    Computers&Accessories|Accessories&Peripherals|...        2477.500000
2    Computers&Accessories|Accessories&Peripherals|...         350.000000
3    Computers&Accessories|Accessories&Peripherals|...         675.500000
4    Computers&Accessories|Accessories&Peripherals|...         650.500000
..                                                 ...                ...
206  OfficeProducts|OfficePaperProducts|Paper|Stati...           0.000000
207  OfficeProducts|OfficePaperProducts|Paper|Stati...           5.000000
208  OfficeProducts|OfficePaperProducts|Paper|Stati...          76.000000
209  OfficeProducts|OfficePaperProducts|Paper|Stati...          52.333333
210  Toys&Games|Arts&Crafts|Drawing&PaintingSupplie...           0.000000

[211 rows x 2 columns]

In [478]:
# Calculate total rating count per category
df.groupby('category')['rating_count'].sum().reset_index()

category  rating_count
0    Car&Motorbike|CarAccessories|InteriorAccessori...          1118
1    Computers&Accessories|Accessories&Peripherals|...          5966
2    Computers&Accessories|Accessories&Peripherals|...          7222
3    Computers&Accessories|Accessories&Peripherals|...         17773
4    Computers&Accessories|Accessories&Peripherals|...         12547
..                                                 ...           ...
206  OfficeProducts|OfficePaperProducts|Paper|Stati...          8420
207  OfficeProducts|OfficePaperProducts|Paper|Stati...          8125
208  OfficeProducts|OfficePaperProducts|Paper|Stati...          8169
209  OfficeProducts|OfficePaperProducts|Paper|Stati...         10487
210  Toys&Games|Arts&Crafts|Drawing&PaintingSupplie...         15867

[211 rows x 2 columns]

#### Aggregation and Group-Based Analysis

GroupBy operations were applied to compute summary statistics such as average discount amount, average rating, and rating count across different categories. This helps identify patterns and trends at the category level.


## Pivot Table

In [479]:
# Create a pivot table to calculate the mean rating for each category
pivot = pd.pivot_table(
    df,                # The DataFrame to pivot
    values='rating',   # The column whose values we want to aggregate
    index='category',  # The column to group by (rows of the pivot table)
    aggfunc='mean'     # Aggregation function: calculate the mean of ratings per category
)
pivot

rating
category                                                    
Car&Motorbike|CarAccessories|InteriorAccessorie...  3.800000
Computers&Accessories|Accessories&Peripherals|A...  4.150000
Computers&Accessories|Accessories&Peripherals|A...  3.500000
Computers&Accessories|Accessories&Peripherals|A...  3.600000
Computers&Accessories|Accessories&Peripherals|A...  4.050000
...                                                      ...
OfficeProducts|OfficePaperProducts|Paper|Statio...  4.250000
OfficeProducts|OfficePaperProducts|Paper|Statio...  4.150000
OfficeProducts|OfficePaperProducts|Paper|Statio...  4.300000
OfficeProducts|OfficePaperProducts|Paper|Statio...  4.133333
Toys&Games|Arts&Crafts|Drawing&PaintingSupplies...  4.300000

[211 rows x 1 columns]

## Cross Tabulation

In [480]:
# Create a crosstab to count occurrences of high rating flags for each category
crosstab = pd.crosstab(
    df['category'],          # Rows: unique values from the 'category' column
    df['high_rating_flag']   # Columns: unique values from the 'high_rating_flag' column
)
crosstab

high_rating_flag                                    0  1
category                                                
Car&Motorbike|CarAccessories|InteriorAccessorie...  1  0
Computers&Accessories|Accessories&Peripherals|A...  0  2
Computers&Accessories|Accessories&Peripherals|A...  1  0
Computers&Accessories|Accessories&Peripherals|A...  2  0
Computers&Accessories|Accessories&Peripherals|A...  0  2
...                                                .. ..
OfficeProducts|OfficePaperProducts|Paper|Statio...  0  2
OfficeProducts|OfficePaperProducts|Paper|Statio...  0  2
OfficeProducts|OfficePaperProducts|Paper|Statio...  0  2
OfficeProducts|OfficePaperProducts|Paper|Statio...  0  3
Toys&Games|Arts&Crafts|Drawing&PaintingSupplies...  0  1

[211 rows x 2 columns]

#### Pivot Tables and Cross-Tabulation

Pivot tables and cross-tabulations were created to analyze relationships between pricing, ratings, and categories. These summaries support multidimensional analysis and are useful for visualization and reporting.




## Encoding

In [ ]:
if 'main_category' not in df.columns:
    df['main_category'] = df['category'].str.split('|').str[0]

# optional sanity check
print(df[['category', 'main_category']].head())

                                            category          main_category
0  Computers&Accessories|Accessories&Peripherals|...  Computers&Accessories
1  Computers&Accessories|Accessories&Peripherals|...  Computers&Accessories
2  Computers&Accessories|Accessories&Peripherals|...  Computers&Accessories
3  Computers&Accessories|Accessories&Peripherals|...  Computers&Accessories
4  Computers&Accessories|Accessories&Peripherals|...  Computers&Accessories


In [482]:
from sklearn.preprocessing import LabelEncoder

df_encoded = df.copy()

le_main = LabelEncoder()
df_encoded['main_category_encoded'] = le_main.fit_transform(
    df_encoded['main_category'].astype(str)
)

le_cat = LabelEncoder()
df_encoded['category_encoded'] = le_cat.fit_transform(
    df_encoded['category'].astype(str)
)

df_encoded[['main_category', 'main_category_encoded']].head()


main_category  main_category_encoded
0  Computers&Accessories                      1
1  Computers&Accessories                      1
2  Computers&Accessories                      1
3  Computers&Accessories                      1
4  Computers&Accessories                      1

# Encoding Techniques Applied in the Project

In the preprocessing of the Amazon product dataset, we applied categorical encoding to transform text-based category columns into numerical formats. This step is essential for enabling machine learning models, correlation analysis, and efficient data processing while maintaining the dataset's usability for visualization and exploration.

### 1. Label Encoding (Primary Method Used)

We applied **Label Encoding** (using `sklearn.preprocessing.LabelEncoder`) to the following columns:

- `main_category` (extracted from the first part of the `category` hierarchy) → encoded as `main_category_encoded`
- `category` (the full hierarchical category path) → encoded as `category_encoded`

**How it works:**  
Each unique category value is assigned a unique integer (e.g., "Computers&Accessories" → 1, "Home&Kitchen" → 3, etc.).

**The Goal:**
- Handle high-cardinality categorical features efficiently (the full `category` column has 211 unique values).
- Provide a compact numerical representation suitable for tree-based models (e.g., Random Forest, XGBoost) and other algorithms that do not assume ordinal relationships between categories.
- Minimize memory usage and avoid excessive feature expansion, making the dataset more scalable.

### 2. One-Hot Encoding (Not Applied)

We deliberately chose **not** to use One-Hot Encoding in this project.

**Reason:**  
The categorical columns have moderate to high cardinality (9 main categories and 211 full subcategories). One-Hot Encoding would create hundreds of new binary columns, leading to:
- Significant increase in dataset dimensionality.
- Higher memory consumption.
- Risk of the "curse of dimensionality" in downstream modeling.

**Chosen Alternative:**  
Label Encoding offers the best trade-off for this dataset, especially given the focus on exploratory data analysis (EDA), visualization, and preparation for models that perform well with label-encoded categories.

---

### 3. Preservation & Result

- **Safe Copy:** All encoding operations were performed on a separate dataframe named `df_encoded`. This ensures the original `df` remains untouched and fully human-readable for reporting, charting, and interpretation.
- **Shape Comparison:**
  - Original `df`: (1465 rows, original columns)
  - `df_encoded`: Same number of rows, with **only +2 additional columns** (`main_category_encoded` and `category_encoded`)
  - **Result:** Efficient numerical

## Conclusion

After applying data cleaning, missing value handling, feature engineering, categorical encoding, outlier detection, and aggregation techniques, the dataset is now clean and well-structured. The processed dataset is suitable for exploratory data analysis, visualization, and predictive analytics.
